### PR TITLE
(Optim): Refactor inputs_merger

### PR DESCRIPTION
**Pull request by AdarshGuptaa**
## 📝 Description
Optimization: inputs_merger Refactor
- Removed Iterative Logic: Switched from a slow, token-by-token CPU loop
- Deleted merged_embeds
- uses Tensor::narrow for zero-copy slicing and a single Tensor::cat for batched GPU concatenation.
- moved core logic into a static function merge_tensors_impl for tests
- Added unit tests

**Fixes/Relates to:** #679
---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** (List new/updated tests)
- [x] **Manual Verification:** (Describe the steps you took)
- [ ] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## Benchmarking
Tested using Criterion benchmark comparing the legacy iterative approach against the new chunked implementation.
Scenario: Realistic 2048-token VLM workload with interleaved images
```
Benchmarking Inputs Merger/Iterative (Stack): Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.2s, enable flat sampling, or reduce sample count to 60.
Inputs Merger/Iterative (Stack)
                        time:   [993.85 µs 996.16 µs 998.92 µs]
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
Inputs Merger/Chunked (Narrow+Cat)
                        time:   [358.58 µs 359.21 µs 359.82 µs]
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  5 (5.00%) high severe
```

~3x increase in efficiency

### CODE REVIEW

**Comment by sidd-27** [comment]
can you also add unit tests for this operation?

**Comment by AdarshGuptaa** [comment]
Added unit tests and made some corrections. Could I get a review?

**Comment by sidd-27** [comment]
Hmm yhea, try the where + scatter approach if possible that might be faster

**Comment by AdarshGuptaa** [comment]
Yessir, I'll look into this

**Comment by AdarshGuptaa** [comment]
Researched a little,
The Scatter approach is not possible because:
- The Mask is a boolean mask instead of a mask of indices

So either we change the mask input type to a tensor of indices or

I'm thinking we can :
1. Using cumsum() function get a culminative sum tensor
2.  we can find the indices where image_embeddings should come by checking the change in sum and use index_select() to essentially stretch the image_embeds across the text_embedding tensor
3. We can verify the streched image_embeds using the mask tensor
4. Finally, using the where_cond() function, we can merge the image_embeds and text_embeds

What do you think?

**Comment by AdarshGuptaa** [comment]
```
fn merge_tensors_impl(
        image_token_mask: &Tensor,
        image_hidden_states: &Tensor,
        inputs_embeds: &Tensor,
    ) -> Result<Tensor> {
        let device = image_token_mask.device();

        // Type conversions for GPU memory savings, cumsum() handling and - 1 handling
        let mask_f32 = image_token_mask.to_dtype(candle_core::DType::F32)?;
        let cumsum = mask_f32.cumsum(0)?;

        // clamp() protects from -1 index underflow and max index overflows
        let one = Tensor::new(&[1.0f32], device)?;
        let max_index = (image_hidden_states.dim(0)? as f64) - 1.0;

        let image_indices = cumsum
            .broadcast_sub(&one)?
            .clamp(0.0, max_index)?
            .to_dtype(candle_core::DType::U32)?;

        // Stretching the image embeddings
        let image_stretched = image_hidden_states.index_select(&image_indices, 0)?;

        // Merging based on special boolean tensor of the mask
        let mask_bool = image_token_mask.ne(0.0)?;
        let mask_broadcast = mask_bool
            .unsqueeze(1)?
            .broadcast_as(inputs_embeds.shape())?;

        let merged_embeds = mask_broadcast.where_cond(&image_stretched, inputs_embeds)?;

        Ok(merged_embeds)
    }
```

Passes Tests, Gotta Benchmark though.

**Comment by AdarshGuptaa** [comment]
```
Modality Merger/Old (Loop & Slice)
                        time:   [425.12 µs 429.45 µs 434.89 µs]
                        change: [-0.41% +0.12% +0.65%] (p = 0.72 > 0.05)
                        No change in performance detected.

Modality Merger/New (Vectorized)
                        time:   [52.120 µs 55.450 µs 59.210 µs]
                        change: [-86.24% -87.10% -87.89%] (p = 0.00 < 0.05)
                        Performance has improved.
```

A 7x-8x Increase on a RTX 3050 6gb laptop gpu

**Comment by sidd-27** [comment]
Lgtm

**Comment by AdarshGuptaa** [comment]
Could I get a review on this @sidd-27 ?

**Comment by AdarshGuptaa** [comment]
This code is not at all good for very large iteration sizes. 

> 32k tokens: crashes my 6gb vram completely

> 16k tokens ~ 390MB (Crash) 

> 8k tokens ~ 190MB (Pass)

**Looked into optimising the approach and came up with a chunk based approach:**
- Slight Overhead but massive **Space Savings**

```
fn merge_tensors_impl(
        image_token_mask: &Tensor,
        image_hidden_states: &Tensor,
        inputs_embeds: &Tensor,
    ) -> Result<Tensor> {
        const BATCH_OR_SEQ_DIM: usize = 0;
        const CHANNEL_OR_EMBED_DIM: usize = 1;
        let (seq_len, _) = inputs_embeds.dims2()?;
        let chunk_size = 1024;
        let mut chunks = Vec::new();

        if image_hidden_states.dim(BATCH_OR_SEQ_DIM)? == 0 {
            return Ok(inputs_embeds.clone());
        }
        let device = image_token_mask.device();

        // Type conversion for GPU memory savings, cumsum() handling and - 1 handling
        let mask_f32 = image_token_mask.to_dtype(candle_core::DType::F32)?;
        let cumsum = mask_f32.cumsum(BATCH_OR_SEQ_DIM)?;

        // clamp() protects from -1 index underflow and max index overflows
        let one = Tensor::new(&[1.0f32], device)?;
        let max_index = (image_hidden_states.dim(BATCH_OR_SEQ_DIM)? as f64) - 1.0;

        let image_indices = cumsum
            .broadcast_sub(&one)?
            .clamp(0.0, max_index)?
            .to_dtype(candle_core::DType::U32)?;
        
        // Chunk based merging
        for start in (0..seq_len).step_by(chunk_size){
            // safety check for final chunk
            let len = usize::min(chunk_size, seq_len-start);

            let mask_chunk = image_token_mask.narrow(0, start, len)?;
            let text_chunk = inputs_embeds.narrow(0, start, len)?;
            let image_indices_chunk = image_indices.narrow(0, start, len)?;

            // Stretch the image embeddings chunk
            let image_chunk_stretched = image_hidden_states.index_select(&image_indices_chunk, BATCH_OR_SEQ_DIM)?;

            // Merging
            let mask_chunk_bool = mask_chunk.ne(0.0)?;
            let mask_chunk_broad = mask_chunk_bool.unsqueeze(CHANNEL_OR_EMBED_DIM)?.broadcast_as(text_chunk.shape())?;
            let merged_chunk = mask_chunk_broad.where_cond(&image_chunk_stretched, &text_chunk)?;
            chunks.push(merged_chunk);
        }

        Tensor::cat(&chunks, 0)
    }
```

Benchmark:
```
Modality Merger/Old (Loop & Slice)
                        time:   [422.15 µs 428.30 µs 435.60 µs]
                        change: [-1.15% +0.35% +1.80%] (p = 0.62 > 0.05)
                        No change in performance detected.

Modality Merger/New (Chunked Vectorized)
                        time:   [64.500 µs 68.120 µs 71.450 µs]
                        change: [-85.12% -84.09% -83.15%] (p = 0.00 < 0.05)
                        Performance has improved.
```

~6x more efficient on an RTX 3050 6gb laptop GPU

I think this is the best balanced implementation between Speed and Space efficiency. What do you think?

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 1]
```diff
@@ -162,3 +194,105 @@ impl SmolModel {
...
+        let (mask, text_embeds, img_embeds) = create_data(vec![1, 1, 1, 1], 4, &Device::Cpu)?;
+
+        let result = SmolModel::merge_tensors_impl(&mask, &img_embeds, &text_embeds)?;
+
+        let result_vec = result.flatten_all()?.to_vec1::<f32>()?;
+
+        assert_eq!(result_vec, vec![2.0, 2.0, 2.0, 2.0]);
+
+        Ok(())
+    }
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: nitpick: too much spacing between the lines?

File: crates/kornia-vlm/src/smolvlm/model.rs
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: better documentation on how this works would be helpful

File: crates/kornia-vlm/src/smolvlm/model.rs (hunk 1 above)
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs]: removed the white spaces

File: crates/kornia-vlm/src/smolvlm/model.rs
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs]: Added Docs and Comments as comprehensively as possible

**Comment by AdarshGuptaa** [comment]
@sidd-27 review?

**Comment by edgarriba** [comment]
LGTM — @sidd-27 can you merge? Thanks!

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 2]
```diff
@@ -162,3 +257,78 @@ impl SmolModel {
...
+
+    #[test]
+    fn test_all_images() -> Result<()> {
+        let (mask, text_embeds, img_embeds) = create_data(vec![1, 1, 1, 1], 4, &Device::Cpu)?;
+        let result = SmolModel::merge_tensors_impl(&mask, &img_embeds, &text_embeds)?;
+        let result_vec = result.flatten_all()?.to_vec1::<f32>()?;
+        assert_eq!(result_vec, vec![2.0, 2.0, 2.0, 2.0]);
+        Ok(())
+    }
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: this is the same as test_img_only_merge, why are there 2 same tests?

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 3]
```diff
@@ -68,40 +67,136 @@ impl SmolModel {
...
+    ///      text tokens and to ensure all indices fall within the image feature bounds.
+    ///
+    /// 3. **Memory-Efficient Merging (Chunking)**:
+    ///    * Iterates through the sequence in windows of 1024 tokens.
+    ///    * Uses `.narrow()` to create zero-copy views, minimizing intermediate allocations.
+    ///    * Performs a vectorized `.where_cond()` within each chunk to swap text embeddings
+    ///      for visual features where the mask is active.
+    ///
+    /// 4. **Reassembly**: Concatenates all processed chunks into a single contiguous
+    ///    tensor along the sequence dimension.
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: very long doc and agentic looking doc comment

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 4]
```diff
@@ -68,40 +67,136 @@ impl SmolModel {
...
+    /// * `inputs_embeds` - The base text embeddings from the language model.
     fn inputs_merger(
         &mut self,
         image_token_mask: &Tensor,
         image_hidden_states: &Tensor,
         inputs_embeds: &Tensor,
     ) -> Result<Tensor> {
-        let total_length = image_token_mask.dims1()?;
+        Self::merge_tensors_impl(image_token_mask, image_hidden_states, inputs_embeds)
+    }
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: why not use the merge_tensors_impl function directly?

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 5]
```diff
@@ -68,40 +67,136 @@ impl SmolModel {
...
+        let cumsum = mask_f32.cumsum(BATCH_OR_SEQ_DIM)?;
+
+        let one = Tensor::new(&[1.0f32], device)?;
+        let max_index = (image_hidden_states.dim(BATCH_OR_SEQ_DIM)? as f64) - 1.0;
+
+        // Align to 0-based indexing by subtracting 1.
+        // Clamp prevents underflow (-1) for leading text tokens and overflow at the tail.
+        let image_indices = cumsum
+            .broadcast_sub(&one)?
+            .clamp(0.0, max_index)?
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: **Nit:** The `.clamp()` here will silently reuse the last image feature if the mask's image token count doesn't match `image_hidden_states.dim(0)`. The old code would panic on out-of-bounds, which is preferablem a mismatch here means a bug upstream, and silent wrong embeddings are harder to debug than a crash.

**Review by sidd-27 — CHANGES_REQUESTED** [review]
overall the comments etc are very agentic and unnecessary, remove a lot of those

File: crates/kornia-vlm/src/smolvlm/model.rs (hunk 2 above)
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs]: removed one

File: crates/kornia-vlm/src/smolvlm/model.rs (hunk 3 above)
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs]: Decreased some bulk

File: crates/kornia-vlm/src/smolvlm/model.rs (hunk 4 above)
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs]: Consolidated the `merge_tensors_impl` into `inputs_merger` and also removed the `&mut self` flag

Updated the `forward` fn to match

File: crates/kornia-vlm/src/smolvlm/model.rs (hunk 5 above)
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs]: the clamp() now truncates `0.0 `and `f64::MAX` so errors are caught later on

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 6]
```diff
@@ -68,40 +67,86 @@ impl SmolModel {
...
+    /// model to process both modalities in a unified manner.
+    ///
+    /// Maps the `image_token_mask` to corresponding 0-based indices in `image_hidden_states`
+    /// using a clamped cumulative sum. To maintain a constant memory footprint, the sequence
+    /// is processed and reassembled in 1024-token chunks using zero-copy views (`.narrow()`)
+    /// and vectorized conditional selection (`.where_cond()`).
+    ///
+    /// # Complexity
+    /// * Time: `O(N)` where `N` is the total sequence length.
+    /// * Space: `O(C * D)` where `C` is the chunk size and `D` is the embedding dimension.
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs]: all of this still feels unnecessary

File: crates/kornia-vlm/src/smolvlm/model.rs [hunk 7]
```diff
@@ -162,3 +207,69 @@ impl SmolModel {
         Ok(())
     }
 }
+
+#[cfg(test)]
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/model.rs:194]: Could you add a test with >1024 tokens to exercise the chunk boundary? All current tests fit in a single chunk, so the narrow → cat reassembly path is untested. Something like a 2049-length sequence with image tokens spanning the boundary would catch off-by-one issues?

**Review by sidd-27 — CHANGES_REQUESTED** [review]

File: crates/kornia-vlm/src/smolvlm/model.rs (hunk 7 above)
Comment (AdarshGuptaa) [reply crates/kornia-vlm/src/smolvlm/model.rs:194]: Proposing to add these tests:
- Alternate embeddings for 2049 seq length
- Contiguous Spanning boundary (1000 text 50 image 990 text 9 image)
- Out of Bounds image_embeddings error
- Empty image and text embeds error
